cirq/protocols/resolve_parameters.py

```diff
@@ -31,10 +31,22 @@ class SupportsParameterization(Protocol):
 
     @document
     def _is_parameterized_(self: Any) -> bool:
-        """Whether the gate is parameterized by any Symbols that require
-        resolution.  Returns True if the gate has any unresolved Symbols
+        """Whether the object is parameterized by any Symbols that require
+        resolution. Returns True if the object has any unresolved Symbols
         and False otherwise."""
 
+    @document
+    def _parameter_names_(self: Any) -> AbstractSet[str]:
+        """Returns a collection of string names of parameters that require
+        resolution. The collection is empty iff _is_parameterized_ is False.
+        """
+
+    @document
+    def _parameter_symbols_(self, Any) -> AbstractSet[sympy.Symbol]:
+        """Returns a collection of sympy Symbols of parameters that require
+        resolution. The collection is empty iff _is_parameterized_ is False.
+        """
+
```

Comment:
This looks to dissociate parameter names and from the symbos that need to be resolved?  Would a better interface be 
_parameters_
which returns a map from string to symbol (and None if it doesn't require resolution)?

One reason to include parameter names even if they don't currently need resolution is this could be used to automatically populate serializated gate sets names.

Also I think if you did this you could end up using a decorator on classes which could autohandle creating these methods: Those that follow the pattern of having args in __init__ equal to attributes and all are parameters could always to this.

Reply:
The two methods are just to get parameters in the format most convenient for the calling code. Names and symbols are in 1:1 correspondence and it's easy to convert between them, e.g. `symbol.name` <-> `sympy.Symbol(name)` so I'm not sure that returning a mapping here would help all that much. Maybe if there were bi-directional maps in python since different callers might want to convert in different directions.

Reply:
Ah, wait, I think I misunderstood your comment. You're imagining returning something like `{'angle': sympy.Symbol('foo'), 'exponent': None}` to indicate that we have "parameters" `angle` and `exponent`, and a symbol `foo` that needs to be resolved? I see some problems with that sort of approach.

* For purposes of the SupportsParametrization protocol, `foo` is what we mean by "parameter" while `angle` and `exponent` are something else, sort of like slots in which parameters may or may not be put.
* "Slots" like `angle` or `exponent` could actually contain expressions involving many symbols that need to be resolved, so at least we'd need to return a `Mapping[str, Set[sympy.Symbol]]`.
* If analyzing a composite object like a circuit, the "slots" are defined on sub-objects like gates within the circuit, so we refer to such slots without also specifying which gate at which position in the circuit; there are no slots on the circuit itself. The idea of the protocols I'm proposing is to strip away all this structure and just get a "flattened" set of parameters that need to be resolved anywhere within the composite object.